### PR TITLE
sql: properly implement create_type_statements index

### LINKED ISSUE #71414: sql: Internal Executor does not work properly with Virtual Index
**Problem** 
When including the INDEX field when defining virtual schema tables in crdb_internal.go and calling it through a built-in, it causes a panic

**To Reproduce**

1. Apply this diff
```
diff --git a/pkg/sql/crdb_internal.go b/pkg/sql/crdb_internal.go
index 28502a8986..0ab8504bd0 100644
--- a/pkg/sql/crdb_internal.go
+++ b/pkg/sql/crdb_internal.go
@@ -2243,7 +2243,8 @@ CREATE TABLE crdb_internal.create_type_statements (
   descriptor_id      INT,
   descriptor_name    STRING,
   create_statement   STRING,
-  enum_members       STRING[] -- populated only for ENUM types
+  enum_members       STRING[], -- populated only for ENUM types
+INDEX (descriptor_id)
 )
 `,
     populate: func(ctx context.Context, p *planner, db catalog.DatabaseDescriptor, addRow func(...tree.Datum) error) error {
```
2. start a server
2. run `create type greeting as enum ('hi');`
3. run `SHOW CREATE ALL TYPES;`

**Expected behavior**
This should not panic since the enums are being properly populated

**Additional data / screenshots**
* ERROR: [n1,client=127.0.0.1:64058,hostssl,user=demo,intExec=crdb_internal.show_create_all_types] a panic has occurred!
* runtime error: index out of range [0] with length 0
* (1) attached stack trace
*   -- stack trace:
*   | runtime.gopanic
*   |   /usr/local/opt/go/libexec/src/runtime/panic.go:1038
*   | runtime.goPanicIndex
*   |   /usr/local/opt/go/libexec/src/runtime/panic.go:90
*   | github.com/cockroachdb/cockroach/pkg/sql.virtualSchemaTable.getIndex
*   |   /Users/jackwu/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_schema.go:220
*   | github.com/cockroachdb/cockroach/pkg/sql.(*virtualDefEntry).makeConstrainedRowsGenerator.func1
*   |   /Users/jackwu/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_schema.go:628
*   | github.com/cockroachdb/cockroach/pkg/sql.setupGenerator.func3
*   |   /Users/jackwu/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_table.go:120
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
*   |   /Users/jackwu/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:491
*   | runtime.goexit
*   |   /usr/local/opt/go/libexec/src/runtime/asm_amd64.s:1581
* Wraps: (2) runtime error: index out of range [0] with length 0
* Error types: (1) *withstack.withStack (2) runtime.boundsError
*
panic: runtime error: index out of range [0] with length 0 [recovered]
        panic: runtime error: index out of range [0] with length 0

goroutine 7396 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc0009ec090, {0x99ef238, 0xc001288740})
        /Users/jackwu/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:254 +0xaa
panic({0x80e4e40, 0xc001568360})
        /usr/local/opt/go/libexec/src/runtime/panic.go:1038 +0x215
github.com/cockroachdb/cockroach/pkg/sql.virtualSchemaTable.getIndex(...)
        /Users/jackwu/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_schema.go:220
github.com/cockroachdb/cockroach/pkg/sql.(*virtualDefEntry).makeConstrainedRowsGenerator.func1({0x99555e0, 0xc000674400})
        /Users/jackwu/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_schema.go:628 +0x7f9
github.com/cockroachdb/cockroach/pkg/sql.setupGenerator.func3({0x99ef238, 0xc001288740})
        /Users/jackwu/go/src/github.com/cockroachdb/cockroach/pkg/sql/virtual_table.go:120 +0x130
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
        /Users/jackwu/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:491 +0x12a
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
        /Users/jackwu/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:481 +0x445

**Environment:**
 - CockroachDB version [e.g. 2.0.x]
 - Server OS: [e.g. Linux/Distrib]
 - Client app [e.g. `cockroach sql`, JDBC, ...]

**Additional context**
What was the impact?

### PR DESCRIPTION
**Pull request by otan**
See individual commits for details!

Resolves #71414

### CODE REVIEW

File: pkg/sql/logictest/testdata/logic_test/crdb_internal [hunk 1]
```diff
@@ -760,6 +760,15 @@ SELECT * FROM crdb_internal.create_type_statements
...
+query ITTITTT
+SELECT * FROM crdb_internal.create_type_statements WHERE descriptor_id = (('enum1'::regtype::oid::int) - 100000)::oid
+----
+54  test  public  66  enum1  CREATE TYPE public.enum1 AS ENUM ('hello', 'hi')  {hello,hi}
+
+query ITTITTT
+SELECT * FROM crdb_internal.create_type_statements WHERE descriptor_id = 'foo'::regclass::oid
+----
+
 # Test the virtual index as well.
```
Comment (rafiss) [inline pkg/sql/logictest/testdata/logic_test/crdb_internal:772]: my fav part of this is the pre-existing comment here :P

File: pkg/sql/crdb_internal.go [hunk 2]
```diff
@@ -2243,53 +2291,60 @@ CREATE TABLE crdb_internal.create_type_statements (
...
-					if err := enumLabelsDatum.Append(tree.NewDString(rep)); err != nil {
-						return err
-					}
+			_, err := writeCreateTypeDescRow(db, sc, typeDesc, addRow)
+			return err
+		})
+	},
+	indexes: []virtualIndex{
+		{
+			populate: func(
```
Comment (rafiss) [inline pkg/sql/crdb_internal.go:2306]: can this share code with the index on pg_catalog.pgtype?

File: pkg/sql/logictest/testdata/logic_test/crdb_internal (hunk 1 above)
Comment (otan) [reply pkg/sql/logictest/testdata/logic_test/crdb_internal:772]: ooh moved this yp!

File: pkg/sql/crdb_internal.go (hunk 2 above)
Comment (otan) [reply pkg/sql/crdb_internal.go:2306]: ok

**Review by rafiss — APPROVED** [review]
thank you!

> Reviewed 3 of 4 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73635#-:-MqUYBXTCb2m3pQpanIY:bsv417h)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

**Comment by otan** [comment]
bors r=rafiss

**Pull request merged**